algorithms/ukrainian.sbl

```diff
@@ -0,0 +1,371 @@
+stringescapes {}
+
+/* the 33 Ukrainian letters and apostrophe represented by single quote */
+
+stringdef a    '{U+0430}'
+stringdef b    '{U+0431}'
+stringdef v    '{U+0432}'
+stringdef gh   '{U+0433}'
+stringdef g    '{U+0491}'
+stringdef d    '{U+0434}'
+stringdef e    '{U+0435}'
+stringdef ye   '{U+0454}'
+stringdef zh   '{U+0436}'
+stringdef z    '{U+0437}'
+stringdef y    '{U+0438}'
+stringdef i    '{U+0456}'
+stringdef yi   '{U+0457}'
+stringdef i`   '{U+0439}'
+stringdef k    '{U+043A}'
+stringdef l    '{U+043B}'
+stringdef m    '{U+043C}'
+stringdef n    '{U+043D}'
+stringdef o    '{U+043E}'
+stringdef p    '{U+043F}'
+stringdef r    '{U+0440}'
+stringdef s    '{U+0441}'
+stringdef t    '{U+0442}'
+stringdef u    '{U+0443}'
+stringdef f    '{U+0444}'
+stringdef kh   '{U+0445}'
+stringdef ts   '{U+0446}'
+stringdef ch   '{U+0447}'
+stringdef sh   '{U+0448}'
+stringdef shch '{U+0449}'
+stringdef soft '{U+044C}'
+stringdef iu   '{U+044E}'
+stringdef ia   '{U+044F}'
+
+// Apostrophe-like symbols
+// stringdef a_apostrophe      '{U+0027}' // '
+// stringdef a_grave_accent   U+0060   // ` cannot to remove system char in Snowball
+stringdef a_ml_prime        '{U+02B9}' // {a_ml_prime}
+stringdef a_mlt_comma       '{U+02BB}' // {a_mlt_comma}
+stringdef a_ml_apostrophe   '{U+02BC}' // {a_ml_apostrophe}
+stringdef a_mlr_comma       '{U+02BD}' // {a_mlr_comma}
+stringdef a_mlv_line        '{U+02C8}' // {a_mlv_line}
+stringdef a_lsq_mark        '{U+2018}' // {a_lsq_mark}
+stringdef a_rsq_mark        '{U+2019}' // {a_rsq_mark}
+stringdef a_shr9q_mark      '{U+201B}' // {a_shr9q_mark}
+stringdef a_prime           '{U+2032}' // {a_prime}
+
+routines (
+  prelude
+  mark_regions R2
+  min_len
+  short_word_4l
+  short_word_5l
+  long_word
+  perfective_gerund
+  reflexive
+  long_endings
+  adjective
+  adjectival
+  verb
+  noun
+  remove_last_letter
+  remove_last_2_letters
+  remove_last_2_vowels
+  remove_vowel_before_vowel
+  remove_last_vowel
+  derivational
+  tidy_up
+)
+
+externals ( stem )
+
+integers ( pV p2 )
+
+groupings ( v )
+
+define v '{a}{e}{ye}{y}{i}{yi}{o}{u}{iu}{ia}'
+
+define prelude as (
+  do repeat ( goto (['{g}']) <- '{gh}' )
+
+  // Remove any apostrophe-like symbols
+  do repeat ( goto (['{'}']) delete )
+  do repeat ( goto (['{a_ml_prime}']) delete )
+  do repeat ( goto (['{a_mlt_comma}']) delete )
+  do repeat ( goto (['{a_ml_apostrophe}']) delete )
+  do repeat ( goto (['{a_mlr_comma}']) delete )
+  do repeat ( goto (['{a_mlv_line}']) delete )
+  do repeat ( goto (['{a_lsq_mark}']) delete )
+  do repeat ( goto (['{a_rsq_mark}']) delete )
+  do repeat ( goto (['{a_shr9q_mark}']) delete )
+  do repeat ( goto (['{a_prime}']) delete )
```

Comment:
Do all these actually occur in real-world Ukrainian text in place of an apostrophe?  There's an overhead to checking for them so I'm dubious about handling characters just because they look kind of like an apostrophe if they don't actually get used in practice.

Possibly snowball should have a more efficient way to transliterate (or delete) a set of characters from in string, but currently the above is a reasonable approach but involves scanning the input once for each character.

Reply:
Until about 10 years ago, there was a lack of Ukrainian keyboard layouts with proper apostrophes and also a lack of OCR software that supported Ukrainian symbols correctly. That resulted in a huge amount of texts, where lots of different Unicode characters that look similar to the apostrophe were used.
In the last decade though the situation improved quite a bit, so now it's mostly down to 3: U+0027, U+02BC, U+2019

Reply:
So I guess it might matter for some cases (i.e. users with a lot of textual data created by OCR over 10 years ago which they've not managed to clean up).

I'm happy for people familiar with the situation to decide what's appropriate - mostly I just wanted to flag this in case this was a instance of attempting theoretical completeness without realising it would add overhead.

Reply:
Well the situation got much better with texts lately. Also with old/unreliable sources I'd expect some text cleaning to happen before they'll be used anywhere anyway. I don't have a strong feeling either way, but if I had to choose I'd say those 3 should be enough for most cases (maybe adding a note to the stemmer's README).